### PR TITLE
HRCPP-282 # Async API and tests

### PR DESCRIPTION
**Pull request by rigazilla**
Async API implementation plus some extra lines to wrap java tests correctly, plus fix for forceReturnValue handling.

https://issues.jboss.org/browse/HRCPP-282 (feature)
https://issues.jboss.org/browse/HRCPP-312 (forceReturnValue bug)

### CODE REVIEW

File: java_tests/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java [hunk 1]
```diff
@@ -352,9 +333,155 @@ public void clear() {
         } catch (cli.Infinispan.HotRod.Exceptions.HotRodClientException ex) {
             throw new HotRodClientException(ex.get_Message(), cause(ex.get_Message()));
         }
-    
     }
 
+   @Override
+   public CompletableFuture<V> putAsync(K k, V v) {
+        try {
+            if (false) workaroundCLICheckedExceptions();
```
Comment (mgencur) [inline java_tests/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java]: If false? That will never happen.

Comment (mgencur) [inline java_tests/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java]: It's a detail but the formatting could be better.

File: src/main/cs/Infinispan/HotRod/Impl/RemoteCacheSWIGImpl.cs [hunk 2]
```diff
@@ -253,6 +261,75 @@ public byte[] Execute(string scriptName, IDictionary<string, string> dict)
             return ret;
         }
 
+        public Task<V> GetAsync(K key)
+        {
+            return Task.Run(() => Get(key));
+        }
+
+        public Task<V> PutAsync(K key, V val, ulong lifespan = 0, TimeUnit lifespanUnit = TimeUnit.SECONDS, ulong maxIdleTime = 0, TimeUnit maxIdleUnit = TimeUnit.SECONDS)
```
Comment (mgencur) [inline src/main/cs/Infinispan/HotRod/Impl/RemoteCacheSWIGImpl.cs]: I believe GetAsync and PutAsync methods should have "async" modifier and should "await" for the Task.Run? ...."The marked async method can itself be awaited by methods that call it."

File: src/main/cs/Infinispan/HotRod/Impl/RemoteCacheSWIGImpl.cs [hunk 3]
```diff
@@ -253,6 +261,75 @@ public byte[] Execute(string scriptName, IDictionary<string, string> dict)
...
+        public Task<V> ReplaceAsync(K key, V val, ulong lifespan = 0, TimeUnit lifespanUnit = TimeUnit.SECONDS, ulong maxIdleTime = 0, TimeUnit maxIdleUnit = TimeUnit.SECONDS)
+        {
+            if (!manager.IsStarted())
+            {
+                String message = "Cannot perform operations on a cache associated with an unstarted RemoteCacheManager. Use RemoteCacheManager.start before using the remote cache.";
+                throw new Infinispan.HotRod.Exceptions.RemoteCacheManagerNotStartedException(message);
+            }
+            else
+            {
+                return Task.Run(() => Replace(key, val, lifespan, lifespanUnit, maxIdleTime, maxIdleUnit));
```
Comment (mgencur) [inline src/main/cs/Infinispan/HotRod/Impl/RemoteCacheSWIGImpl.cs]: Again. "async" and "await" ?

File: src/main/cs/Infinispan/HotRod/Impl/RemoteCacheSWIGImpl.cs [hunk 4]
```diff
@@ -253,6 +261,75 @@ public byte[] Execute(string scriptName, IDictionary<string, string> dict)
...
+        }
+
+        public Task<bool> ReplaceWithVersionAsync(K key, V val, ulong version, ulong lifespan = 0, ulong maxIdleTime = 0)
+        {
+            if (!manager.IsStarted())
+            {
+                String message = "Cannot perform operations on a cache associated with an unstarted RemoteCacheManager. Use RemoteCacheManager.start before using the remote cache.";
+                throw new Infinispan.HotRod.Exceptions.RemoteCacheManagerNotStartedException(message);
+            }
+            return Task.Run(() => ReplaceWithVersion(key, val, version, lifespan, maxIdleTime));
```
Comment (mgencur) [inline src/main/cs/Infinispan/HotRod/Impl/RemoteCacheSWIGImpl.cs]: Again. "async" and "await" ?

File: src/main/cs/Infinispan/HotRod/Impl/RemoteCacheSWIGImpl.cs [hunk 5]
```diff
@@ -253,6 +261,75 @@ public byte[] Execute(string scriptName, IDictionary<string, string> dict)
...
+        public Task<bool> RemoveWithVersionAsync(K key, ulong version)
+        {
+            if (!manager.IsStarted())
+            {
+                String message = "Cannot perform operations on a cache associated with an unstarted RemoteCacheManager. Use RemoteCacheManager.start before using the remote cache.";
+                throw new Infinispan.HotRod.Exceptions.RemoteCacheManagerNotStartedException(message);
+            }
+            else
+            {
+                return Task.Run(() => RemoveWithVersion(key, version));
```
Comment (mgencur) [inline src/main/cs/Infinispan/HotRod/Impl/RemoteCacheSWIGImpl.cs]: Again. "async" and "await" ?

File: java_tests/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java (hunk 1 above)
Comment (rigazilla) [inline java_tests/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java]: ok, done

File: java_tests/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java [hunk 6]
```diff
@@ -352,9 +333,155 @@ public void clear() {
...
+          Boolean res= jniRemoteCache.taskResultBool(t);
+          return res;
+      });
+        } catch (cli.Infinispan.HotRod.Exceptions.RemoteCacheManagerNotStartedException ex) {
+            throw new RemoteCacheManagerNotStartedException(ex.get_Message());
+        } catch (cli.Infinispan.HotRod.Exceptions.HotRodClientException ex) {
+            throw new HotRodClientException(ex.get_Message(), cause(ex.get_Message()));
+        }
+   }
+
```
Comment (rigazilla) [inline java_tests/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java]: ok, done

File: src/main/cs/Infinispan/HotRod/IRemoteCache.cs [hunk 7]
```diff
@@ -1,392 +1,488 @@
...
+        ///   Async version of the RemoveWithVersion method
+        /// </summary>
+        ///
+        /// <param name="key">key</param>
+        /// <param name="version">version numeric version that should match the one in the server for the operation to succeed</param>
+        ///
+        /// <returns>An instance of Task&lt;bool&gt; that will contains the result of the Remove</returns>
+        /// <see cref="RemoveWithVersion(K, ulong)"/>
+        Task<bool> RemoveWithVersionAsync(K key, ulong version);
+
```
Comment (rigazilla) [inline src/main/cs/Infinispan/HotRod/IRemoteCache.cs]: Ok, done

File: src/main/cs/Infinispan/HotRod/Impl/RemoteCacheSWIGImpl.cs (hunk 3 above)
Comment (rigazilla) [inline src/main/cs/Infinispan/HotRod/Impl/RemoteCacheSWIGImpl.cs]: If the method is marked with async it cannot raise exceptions and this leads to test failure (e.g. CacheManagerNotStartedTest.testPutAsync()). With the current implementation the Async methods just execute the operation in a parallell task and can't block so there nothing to await.

File: java_tests/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java [hunk 8]
```diff
@@ -354,6 +354,72 @@ public void clear() {
...
+       for (Entry<? extends K, ? extends V> entry : m.entrySet()) {
+           dict.Add(marshal(entry.getKey()), marshal(entry.getValue()));
+       }
+       return dict;
+   }
+   
+   @Override
+   public CompletableFuture<Void> putAllAsync(Map<? extends K, ? extends V> map) {
+       try {
+           if (false) workaroundCLICheckedExceptions();
```
Comment (mgencur) [inline java_tests/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java]: I don't understand this "if (false) ...." Will this ever be called? What's the purpose?

Comment (rigazilla) [inline java_tests/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java]: Uh sorry, I didn't reply to this. It's an old piece of code, I'm not sure if there's actually a workaround behind, maybe for IKVM. I hadn't time to try a build without these lines. We can leaves these as they are just for now? I'll clean up the code later.

Comment (mgencur) [inline java_tests/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java]: OK. We can remove it a little later. Thanks.

**Comment by mgencur** [comment]
This PR looks good to me now.

**Comment by mgencur** [comment]
Pulling.

**Comment by mgencur** [comment]
Pushed. Thanks.

**Pull request closed without merging**